schema/query/projection_evaluator.go

```diff
@@ -42,7 +42,7 @@ func (p Projection) Eval(ctx context.Context, payload map[string]interface{}, rs
 func evalProjection(ctx context.Context, p Projection, payload map[string]interface{}, validator schema.Validator, rbr *referenceBatchResolver) (map[string]interface{}, error) {
 	res := map[string]interface{}{}
 	resMu := sync.Mutex{} // XXX use sync.Map once Go 1.9 is out
-	if len(p) == 0 {
+	if len(p) == 0 || p[0].Name == "*" {
```

Comment:
What if * is supplied with other fields? Shouldn't we reject it?

Reply:
I am not sure. Having `address,*` will have all fields `address` included. I can add a condition, where `*` to be the only predicate, else ignore it, but It doesn't feel right.

Reply:
@smyrman can you comment on this as well.

Reply:
I am not completely sure I follow your comment @rs.

I thought the whole benefit of the syntax was that you could do `fields="*,sub1{},sub2{},hiddenField"`, to expand `sub1` and `sub2` and maybe explicitly include `hiddenField`, while still including all fields that where not explicitly mentioned.

Am I missing something?

Reply:
Or do you mean support `address,*` as well as `*,address`?

Reply:
I thought about `sub{*}` but not `*,sub{}`. The problem I think we have here is that `*` will already contain `sub` as an id. So we will end up with a duplicated field.

Did I miss something?

Reply:
@smyrman Having sub-resource like `sub{}` is not possible at the moment. My patch will allow `sub{*}` which is equivalent of just saying `sub`

Reply:
@smyrman 
`address,*` and `*,address` will result the same result as just `*`. I guess @rs was implying that we should rise an error on `*,address`/`address,*` cases, where I think they are completely valid.

Reply:
@Dragomir-Ivanov I meant `*,sub{<something>}`. My point is to say that you patch might not allow `*` with any other non aliased field as it would result in duplicate field definition. I don't think we check that anyway, but my point is that the result would be unpredictable.

To be more specific, if we have a schema with the field `a` as string and `b` as sub field. Asking for `*,b{*}` would be like asking for `a,b,b{*}`. Will the result contain `b` as an id or `b` as a sub struct is unclear and depend on the implementation.

Reply:
Okay, think I am missing something. Don't understand what you are explaining, and it is because I don't know `rest-layer` that well. Are you talking about nested objects within a resource ( `schema.Object` )

Reply:
Yes

Reply:
Still can't quite understand the purpose of `schema.Object` and `schema.Connection` but I tried the following:
```
			"a": {
				Validator: &schema.String{},
			},
			"b": {
				Schema: &schema.Schema{
					Fields: schema.Fields{
						"a": {
							Validator: &schema.String{},
						},
						"b": {
							Validator: &schema.String{},
						},
					},
				},
			},
```
This is the data in the resource:
```
{
  "a": "Data 1",
  "b": {
    "a": "Data 2",
    "b": "Data 3"
  },
  "created": "2017-10-13T12:38:34.399542387+03:00",
  "id": "b7g8j6h40fm2fhph21ag",
  "name": "EMA",
  "updated": "2017-10-13T12:38:34.399543697+03:00"
}
```
All these requests:
```
GET http://192.168.0.103:3001/api/users/b7g8jpp40fm2fhph21eg?fields=*
GET http://192.168.0.103:3001/api/users/b7g8jpp40fm2fhph21eg?fields=*,b
GET http://192.168.0.103:3001/api/users/b7g8jpp40fm2fhph21eg?fields=*,b{*}
GET http://192.168.0.103:3001/api/users/b7g8jpp40fm2fhph21eg?fields=*,b{*},b{*}
GET http://192.168.0.103:3001/api/users/b7g8jpp40fm2fhph21eg?fields=*,b{*},b{*},*
and even
GET http://192.168.0.103:3001/api/users/b7g8jpp40fm2fhph21eg?fields=*,b{a}     This is because * has precedence  
```
Result the same response:
```
{
  "a": "Data 1",
  "b": {
    "a": "Data 2",
    "b": "Data 3"
  },
  "created": "2017-10-13T12:39:51.670839057+03:00",
  "id": "b7g8jpp40fm2fhph21eg",
  "name": "EMA",
  "updated": "2017-10-13T12:39:51.67084017+03:00"
}
```
Trying something different:
```
GET http://192.168.0.103:3001/api/users/b7g8jpp40fm2fhph21eg?fields=a,b{a}
```
Gets expected result:
```
{
  "a": "Data 1",
  "b": {
    "a": "Data 2"
  }
}
```

Reply:
When I talk abut sub-resources and hidden fields, I was thinking more like this:
```go
resourceA := index.Bind("parent", schema.Schema{
    Fields: {
        "showMe": {},
        "hiddenField": {Hidden: true},
}, ...)
rscParent := index.Bind("Parent", schema.Schema{
    Fields: {
        "showMe": {},
        "hiddenField": {Hidden: true},
    },
}, storeA, conf)
// afterc calling rscA.Bind, "sub" will now appear as a `schema.Connection` in `rscA` on a special _field override list_.
rscA.Bind("sub", "parent", schema.Schema{
    Fields: {
        "parent": {Validator: &schema.Reference{Path:"parent"}},
        "a": {},
        "hiddenField": {Hidden: true},
    },
})
```

And then ask for either `/parent?fields=*,sub{hiddenField}` or /parent?fields=sub{hiddenField},*`

So I guess `@rs` point is, if this is expanded to either `fields=showMe,sub,sub{a}` or  `fields=sub{a},showMe,sub`, which fields will `sub` really show? Clearly our intention of the user would be to display only `hiddenField`, but what will happen?

The bet approach for this discussion is probably to first add some tests for this, and then see what changes is needed for the implementaiton in order for the user assumption to hold.

Reply:
Btw, The test you showed us was really nice @Dragomir-Ivanov. But I am not sure I would _want_ `*` to have precedence...

Reply:
This asks another question too, should `fields=foo,foo` or `fields=f:foo,f:bar` raise an error or just result in a "last win"?

Reply:
@rs Good question, I guess we should just pick one, and keep it until it proves wrong. I vote for `last win`.

Reply:
That’s the current implementation. We should add tests and document it then.